File: go/day8/part2/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	fmt.Println("AoC Day 7 Part 2")

	if len(os.Args) < 2 {
		fmt.Println("Need to provide input")
		return
	}

	input := os.Args[1]
	processInput(input)
}

func processInput(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	grid := [][]int{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		row := []int{}
		for _, v := range line {
			height, _ := strconv.Atoi(string(v))
			row = append(row, height)
		}

		grid = append(grid, row)
	}

	scenicScore := 0
	for y, yv := range grid {
		for x, _ := range yv {
			score := checkVis(grid, x, y, false)
			if score > scenicScore {
				scenicScore = score
			}
		}
	}
	fmt.Println(scenicScore)
}

func checkVis(grid [][]int, x, y int, dbg bool) int {
	// Check North
	nScore := 0
	for i := y - 1; i >= 0; i-- {
		nScore++
		if grid[i][x] >= grid[y][x] {
			break
		}
	}
	// Check South
	sScore := 0
	for i := y + 1; i < len(grid); i++ {
		sScore++
		if grid[i][x] >= grid[y][x] {
			break
		}
	}

	// Check East
	eScore := 0
	for i := x + 1; i < len(grid[y]); i++ {
		eScore++
		if grid[y][i] >= grid[y][x] {
			break
		}
	}

	// Check West
	wScore := 0
	for i := x - 1; i >= 0; i-- {
		wScore++
		if grid[y][i] >= grid[y][x] {
			break
		}
	}

	return nScore * wScore * sScore * eScore
}

File: nim/day1/part2.nim
import os
import parseutils

proc push(ar:var array[3,int], val:int):void =
  if val > ar[0]:
    ar[2] = ar[1]
    ar[1] = ar[0]
    ar[0] = val
  elif val > ar[1]:
    ar[2] = ar[1]
    ar[1] = val
  elif val > ar[2]:
    ar[2] = val

proc processInput(filename:string):void =
  let file = open(filename)
  defer: file.close()

  var 
    cal = [0,0,0]
    tmp = 0

  for line in file.lines():
    if line.len < 1:
      tmp = 0
      continue
    
    var res:int
    discard parseInt(line,res,0)
    tmp += res
    
    cal.push(tmp)
  
  echo cal[0]+cal[1]+cal[2]


proc main():void =
  echo "AoC Day 1 Part 2"
  let params = commandLineParams()

  if params.len < 1 :
    echo "Need to provide input"
    return
  
  processInput(params[0])

  
main()